File: Ballast.Core/src/Client/models/tile.ts
import { ICubicCoordinates, CubicCoordinates } from './cubic-coordinates';
import { ITileShape, TileShape } from './tile-shape';
import { ITerrain, Terrain } from './terrain';

export interface ITile {
    cubicCoordinates: ICubicCoordinates;
    tileShape: ITileShape;
    terrain: ITerrain;
}

export class Tile implements ITile {

    public readonly cubicCoordinates: CubicCoordinates;
    public readonly tileShape: TileShape;
    public readonly terrain: Terrain;
    
    private constructor(state: ITile) {
        this.cubicCoordinates = CubicCoordinates.fromObject(state.cubicCoordinates);
        this.tileShape = TileShape.fromObject(state.tileShape);
        this.terrain = Terrain.fromObject(state.terrain);
    }

    public static fromObject(object: ITile): Tile {
        return new Tile(object);
    }

    public static *fromObjectList(objects: ITile[]) : IterableIterator<Tile> {
        if (!objects || objects.length < 1) {
            return [];
        }
        for(let object of objects) {
            yield Tile.fromObject(object);
        }
    }

}

File: Ballast.Core/src/Server/Models/Tile.cs
namespace Ballast.Core.Models
{
    public class Tile : ITile 
    {

        private readonly CubicCoordinates _cubicCoordinates;
        private readonly TileShape _tileShape;
        private readonly Terrain _terrain;
        
        public ICubicCoordinates CubicCoordinates => _cubicCoordinates;
        public ITileShape TileShape => _tileShape;
        public ITerrain Terrain => _terrain;

        private Tile(ICubicCoordinates cubicCoordinates, ITileShape tileShape, ITerrain terrain)
        {
            _cubicCoordinates = Models.CubicCoordinates.FromObject(cubicCoordinates);
            _tileShape = Models.TileShape.FromObject(tileShape);
            _terrain = Models.Terrain.FromObject(terrain);
        }

        private Tile(ITile state) : this(
            cubicCoordinates: state.CubicCoordinates,
            tileShape: state.TileShape,
            terrain: state.Terrain
        ) { }

        public static Tile FromObject(ITile state) => new Tile(state);

        public static Tile FromProperties(ICubicCoordinates cubicCoordinates, ITileShape tileShape, ITerrain terrain) => 
            new Tile(
                cubicCoordinates: cubicCoordinates,
                tileShape: tileShape,
                terrain: terrain
            );

    }
}